File: daily_python/2109_Adding_Spaces_to_a_String.py
import os
import sys
from typing import List
from collections import defaultdict, deque

class Solution:
    def addSpaces(self, s: str, spaces: List[int]) -> str:
        n = len(s)
        m = len(spaces)
        space_idx = 0
        ans = []
        for i in range(n):
            if space_idx < m and i == spaces[space_idx]:
                ans.append(' ')
                space_idx += 1
            ans.append(s[i])
        return "".join(ans)

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            spaces = list(map(int, lines[1].split('=')[1].strip()[1:-2].split(',')))
            dataset.append((s, spaces))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s, spaces in dataset:
        results.append(solution.addSpaces(s, spaces))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/2109_Adding_Spaces_to_a_String.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    string addSpaces(string s, vector<int>& spaces) {
        int n = s.length();
        int m = spaces.size();
        string ans(n + m, ' ');
        int idx = 0;
        int space_idx = 0;
        int ans_idx = 0;
        while(space_idx < m){
            int k = (space_idx == 0)? spaces[space_idx] : spaces[space_idx] - spaces[space_idx - 1];
            space_idx++;
            while(k--){
                ans[ans_idx++] = s[idx++];
            }
            ans_idx++;
        }
        while(idx < n) ans[ans_idx++] = s[idx++];
        return ans;
    }
};

int main(){
    string s = "LeetcodeHelpsMeLearn";
    vector<int> spaces = {8,13,15};
    // string s = "icodeinpython";
    // vector<int> spaces = {1,5,7,9};
    // string s = "spacing";
    // vector<int> spaces = {0,1,2,3,4,5,6};
    Solution S;
    cout << S.addSpaces(s, spaces) << endl;
    return 0;
}